applications/plugins/Communication/components/serverCommunicationVRPN.inl

```diff
@@ -0,0 +1,399 @@
+}
+
+std::string ServerCommunicationVRPN::defaultDataType()
+{
+    return "VRPNstring";
+}
+
+/******************************************************************************
+*                                                                             *
+* SEND PART                                                                   *
+*                                                                             *
+******************************************************************************/
+
+void ServerCommunicationVRPN::sendData()
+{
+    std::string address = d_address.getValueString();
+    m_connection = vrpn_create_server_connection();
+    std::map<std::string, CommunicationSubscriber*> subscribersMap = getSubscribers();
+    if (subscribersMap.size() == 0)
+    {
+        if (isVerbose())
+            msg_info(this) << "Server Communication VRPN does not have Subscriber";
+        return;
+    }
+    for (std::map<std::string, CommunicationSubscriber*>::iterator it = subscribersMap.begin(); it != subscribersMap.end(); it++)
+    {
+        CommunicationSubscriber* subscriber = it->second;
+        std::string strTestText = subscriber->getSubject()+"@"+address;
+        const char *device = strTestText.c_str();
+        vrpn_text_sender = new vrpn_Text_Sender(device, m_connection);
+        vrpn_analog_server = new vrpn_Analog_Server(device, m_connection);
+        vrpn_tracker_server = new vrpn_Tracker_Server(device, m_connection);
+    }
+
+    while (!m_connection->connected() && this->m_running)
+        m_connection->mainloop();
+
+    while (m_connection->connected() && this->m_running)
+    {
+
+        for (std::map<std::string, CommunicationSubscriber*>::iterator it = subscribersMap.begin(); it != subscribersMap.end(); it++)
+        {
+            CommunicationSubscriber* subscriber = it->second;
+            ArgumentList argumentList = subscriber->getArgumentList();
+
+            try
+            {
+                for (ArgumentList::iterator itArgument = argumentList.begin(); itArgument != argumentList.end(); itArgument++ )
+                    createVRPNMessage(subscriber, *itArgument);
+            } catch(const std::exception& e) {
+                if (isVerbose())
+                    msg_info("ServerCommunicationVRPN") << e.what();
+            }
+            m_connection->mainloop(&delay);
+        }
+    }
+}
+
+void ServerCommunicationVRPN::createVRPNMessage(CommunicationSubscriber* subscriber, std::string argument)
+{
+    std::stringstream messageStr;
+    BaseData* data = fetchDataFromSenderBuffer(subscriber, argument);
+    if (!data)
+        throw std::invalid_argument("data is null");
+    const AbstractTypeInfo *typeinfo = data->getValueTypeInfo();
+    const void* valueVoidPtr = data->getValueVoidPtr();
+
+    delay.tv_sec = 0L;
+    delay.tv_usec = 0L;
+    angle += 0.001f;
+
+    if (!typeinfo->Container())
+    {
+        if (vrpn_text_sender)
```

Comment:
we should check the data type instead of using all servers

Reply:
Do you mean we should set up a condition on `receivers` vector for different data types? Then how we will differentiate analog and tracker type?

Reply:
cast ?